File: problems/Diagonal Traverse/findDiagonalOrder.cs
public class Solution {
    public int[] FindDiagonalOrder(int[][] matrix) {
        if (0 == matrix.Length) {
            return new int[]{};
        }

        int rows = matrix.Length;
        int cols = matrix[0].Length;
        List<List<int>> diagonals = new List<List<int>>();
        List<int> result = new List<int>(rows * cols);

        for (int j = 0; rows + cols > j; ++j) {
            int i = 0;
            int k = j;

            if (cols <= j) {
                k = cols - 1;
                i = 1 + j - cols;
            }

            diagonals.Add(new List<int>());

            while (rows > i && 0 <= k) {
                diagonals[diagonals.Count - 1].Add(matrix[i++][k--]);
            }
        }

        for (int i = 0; diagonals.Count > i; ++i) {
            if (0 == (i & 1)) {
                diagonals[i].Reverse();
            }
            result.AddRange(diagonals[i]);
        }

        return result.ToArray();
    }
}

File: problems/Diagonal Traverse/findDiagonalOrder.java
class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int rows = mat.length;
        int cols = mat[0].length;
        int n = rows * cols;
        int[] result = new int[n];
        HashMap<Integer, List<Integer>> diagonalIdxToValuesMap = new HashMap<>();
        
        for (int row = 0; rows > row; ++row) {
            for (int col = 0; cols > col; ++col) {
                if (!diagonalIdxToValuesMap.containsKey(row + col)) {
                    diagonalIdxToValuesMap.put(row + col, new ArrayList<>());
                }

                diagonalIdxToValuesMap.get(row + col).add(mat[row][col]);
            }
        }

        for (int diagonalIdx = 0, idx = 0; rows + cols - 1 > diagonalIdx; ++diagonalIdx) {
            List<Integer> items = diagonalIdxToValuesMap.get(diagonalIdx);
            int nextIdx = idx;

            idx += items.size();

            if (0 == diagonalIdx%2) {
                nextIdx += items.size() - 1;
            }

            for (int item : items) {
                result[nextIdx] = item;

                if (0 == diagonalIdx%2) {
                    --nextIdx;
                } else {
                    ++nextIdx;
                }
            }
        }

        return result;
    }
}
